polaris-core/src/main/java/org/apache/polaris/core/persistence/LocalPolarisMetaStoreManagerFactory.java

```diff
@@ -96,37 +97,40 @@ private void initializeForRealm(
   }
 
   @Override
-  public synchronized Map<String, PrincipalSecretsResult> bootstrapRealms(
-      List<String> realms, PolarisCredentialsBootstrap credentialsBootstrap) {
-    Map<String, PrincipalSecretsResult> results = new HashMap<>();
-
-    for (String realm : realms) {
-      RealmId realmId = RealmId.newRealmId(realm);
-      if (!metaStoreManagerMap.containsKey(realmId.id())) {
-        initializeForRealm(realmId, credentialsBootstrap);
+  public synchronized Map<RealmId, PrincipalSecretsResult> bootstrapRealms(
+      List<RealmId> realms, PolarisCredentialsBootstrap credentialsBootstrap) {
+    Map<RealmId, PrincipalSecretsResult> results = new HashMap<>();
+
+    for (RealmId realm : realms) {
+      if (!metaStoreManagerMap.containsKey(realm.id())) {
+        initializeForRealm(realm, credentialsBootstrap);
         PrincipalSecretsResult secretsResult =
             bootstrapServiceAndCreatePolarisPrincipalForRealm(
-                realmId, metaStoreManagerMap.get(realmId.id()));
-        results.put(realmId.id(), secretsResult);
+                realm, metaStoreManagerMap.get(realm.id()));
+        results.put(ImmutableRealmId.copyOf(realm), secretsResult);
```

Comment:
why copying? it's immutable

Reply:
It's `RequestScoped`, so I think it can get lost.

Reply:
It's potentially a CDI bean, and these are poor choices as map keys: the CDI specs do not say anything about how CDI proxies should handle equals() and hashCode(). Such beans can have surprising effects when put in a set or a map.

Reply:
I worry that this is an extremely easy mistake to make. I dislike proxies for this and other reasons. A person has to have a lot of context in their head before they do simple things like add keys to maps.

Reply:
@eric-maynard @adutra fair point. However, the realm-IDs here are parameters for bootstrap?

Reply:
Unfortunately the `bootstrapRealms` method can be called both from a bootstrap command (in which case `RealmId` is not a CDI proxy) and from "regular" server code (in which case it is a proxy). (The "regular" path only happens with the in-memory metastore though.)